### PR TITLE
Add an example of using automm in kaggle competition pawpularity.

### PR DESCRIPTION
**Pull request by Linuxdex**
*Issue #, if available:*
It is an example of using automm to complete Pawpularity kaggle competition.
*Description of changes:*
Including three files. 
README.md is the instruction document of using automm in kaggle.
kaggle_Pawpularity.py is the training part of the predictor.
kaggle_Pawpularity_submit.py is the submission part of the predictor in kaggle.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+﻿Take Petfinder Pawpularity competition as an example showing how to use AutoMMPredictor to complete a competition.
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:1]: Add a link about Petfinder `[Petfinder Pawpularity](https://www.kaggle.com/competitions/petfinder-pawpularity-score/overview)`. Consider to change to the following:

```
Take [Petfinder Pawpularity](https://www.kaggle.com/competitions/petfinder-pawpularity-score/overview) competition as an example to show how to use AutoMMPredictor to complete in a Kaggle competition.
```

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+## 1. Kaggle Competition with AutoGluon
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:3]: We need to emphasize that it's *kernel-only* competition.

```

## 1. Kaggle Kernel-only Competition with AutoGluon

```

**Comment by szha** [comment]
Job PR-1796-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1796/1/index.html

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+## 1. Kaggle Kernel-only Competition with AutoGluon
+
+In a kaggle competition, especially a code competition, users cannot obtain AutoGluon resources through the network. 
+To solve the problem, there are two key points:
+
+ - Get AutoGluon and its dependent libraries.
+ - Save the downloaded resources when saving your model.
+
+### 1.1 Get AluoGluon in Kaggle
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:11]: For this part, need to to be more clear what should we do to "upload the installed AutoGluon" to Kaggle.

**Comment by szha** [comment]
Job PR-1796-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1796/2/index.html

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ - timm
+ - omegaconf
+ - antlr4
+
+### 1.2 Save Standalone Model
+In AutoMMPredictor, some pretrained models will be downloaded during training. These models also need to be saved for use in predicting after submission. You can open standalone in model saving to save the downloaded models.
+
+    predictor.save(path=save_standalone_path,standalone=True)
+
+## 2. AutoMMPredictor for Training
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:33]: I suggest to move this section above. We first teach the user about how to train the model and then teach them how to submit it in Kaggle.

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+AutoMMPredictor is a tool that can handle diverse data including image, text, numerical data and categorical data. It can automatically identify data types and perform fusion. With the tool, you can make the train easily with less code. For details, please refer to [`kaggle_Pawpularity.py`](./kaggle_Pawpularity.py).
+
+### 2.1 Build the AutoMMPredictor
+
+You can build the predictor as follow.
+
+    predictor = AutoMMPredictor(
+	    label="Pawpularity", 
+	    problem_type="regression", 
+	    eval_metric="rmse",
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:44]: We can potentially mention bce loss in another tutorial.

Comment (sxjscience) [reply examples/automm/Kaggle-Pawpularity/README.md:44]: Oh, I misread the code. Ignore the comment above.

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+ - `hyperparameters`is a Dict which will override the default configs in the training. The configs contain five different types.  
+ -- `model`contains the parameters which control the models used in the predictor. You can select the model you need and adjust the details. Default is selecting the models determined by dataset automatically.
+ --`data`contains the configs of transforms for different types of data. 
+ --`env`contains the configs of the training environment.
+ --`optimization`contains the configs in the optimization process, including but not limit to max training epochs, learning rate and warm up.
+ - `seed`determines the random seed.
+
+## 3. Prediction in Kaggle Competitions
+
+In kaggle competitions, the predictor you trained will be used to predict the test dataset. You should upload the AutoMMPredictor standalone save files as a dataset to kaggle and put it in the input data sources of your prediction code.
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:94]: Next, let's upload the predictor to Kaggle and use it to generate predictions on the test set. You should upload the AutoMMPredictor standalone save files as a dataset to kaggle and put it in the input data sources of your prediction code.

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
+﻿Take [Petfinder Pawpularity competition](https://www.kaggle.com/competitions/petfinder-pawpularity-score/overview) as an example showing how to use AutoMMPredictor to complete a competition.
+
+## 1. AutoMMPredictor for Training
+
+AutoMMPredictor is a tool that can handle diverse data including image, text, numerical data and categorical data. It can automatically identify data types and perform fusion. With the tool, you can make the train easily with less code. For details, please refer to [`kaggle_Pawpularity.py`](./kaggle_Pawpularity.py).
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:5]: It can automatically identify data types and fuse the pretrained DL backbones.

Comment (sxjscience) [reply examples/automm/Kaggle-Pawpularity/README.md:5]: you can train the model easily without writing lots of code.

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    predictor = AutoMMPredictor(
+	    label="Pawpularity", 
+	    problem_type="regression", 
+	    eval_metric="rmse", 
+	    path=save_path,  
+	    verbosity=4, 
+	)
+
+ - `label`indicates the target value in training data.
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:19]: Need to add a space between label and indicates

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+		    "model.timm_image.train_transform_types": "['resize_shorter_side','center_crop','randaug']",
+		    "data.categorical.convert_to_text": "False",
+		    "env.per_gpu_batch_size": "16",
+		    "env.per_gpu_batch_size_evaluation": "32",
+		    "env.precision": "32",
+		    "optimization.learning_rate": "2e-5",
+		    "optimization.weight_decay": "0",
+		    "optimization.lr_decay": "1",
+		    "optimization.max_epochs": "5",
+		    "optimization.warmup_steps": "0",
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:45]: Is warmup useless for this task?

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+ - `save_path`indicates the specific path for model saving in a fit process.
+ - `hyperparameters`is a Dict which will override the default configs in the training. The configs contain five different types.  
+ -- `model`contains the parameters which control the models used in the predictor. You can select the model you need and adjust the details. Default is selecting the models determined by dataset automatically.
+ --`data`contains the configs of transforms for different types of data. 
+ --`env`contains the configs of the training environment.
+ --`optimization`contains the configs in the optimization process, including but not limit to max training epochs, learning rate and warm up.
+ - `seed`determines the random seed.
+
+### 1.3 Save Standalone Model
+In AutoMMPredictor, some pretrained models will be downloaded during training. These models also need to be saved for use in predicting after submission. You can open standalone in model saving to save the downloaded models.
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:62]: you can specify the predictor to save a “standalone” model that can be loaded without internet access.

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    predictor.save(path=save_standalone_path,standalone=True)
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:64]: Add space before “,”.

File: examples/automm/Kaggle-Pawpularity/README.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+## 2. Kaggle Kernel-only Competition with AutoGluon
+
+In a kaggle competition, especially a code competition, users cannot obtain AutoGluon resources through the network. 
+To solve the problem, there are two key points:
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/README.md:69]: Rephrase the “key points”.

**Comment by szha** [comment]
Job PR-1796-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1796/3/index.html

**Comment by szha** [comment]
Job PR-1796-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1796/4/index.html

File: examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity_submit.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            verbosity=4,
+        )
+        pretrained_model = predictor.load(path=save_standalone_path + f'_fold{fold}/')
+        test_pred = pretrained_model.predict(test_df)
+        all_preds_3.append(test_pred)
+        del predictor
+        torch.cuda.empty_cache()
+    preds_3 = np.mean(np.stack(all_preds_3), axis=0)
+
+    save_path = "../input/pawpularity-result/result4/result"
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity_submit.py:88]: Should we change to use a for loop?

Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity_submit.py:88]: Should we change to use a for loop?

**Comment by szha** [comment]
Job PR-1796-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1796/5/index.html

File: examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        # Training process.
+        training_df = train_df[train_df_fold != i]
+        valid_df = train_df[train_df_fold == i]
+        predictor.fit(
+            train_data=training_df,
+            tuning_data=valid_df,
+            save_path=save_path + f"_fold{i}",
+            hyperparameters={
+                # "model.names": "['timm_image']",
+                "model.timm_image.checkpoint_name": "swin_large_patch4_window7_224",
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity.py:73]: Consider to use ArgumentParser so that the user can specify the backbone. You may bring the logic from https://github.com/awslabs/autogluon/blob/aa6e9591b452154b7bc3d2edb897bde271b79cfc/examples/text_prediction/run_competition.py#L12 . Consider to have a key called “--image_backbone” and then have a key called “--fusion”

File: examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    test_df = pd.read_csv(os.path.join(data_path, "test.csv"))
+    test_df.rename(columns={"Id": "Image Path"}, inplace=True)
+    test_df["Image Path"] = test_df["Image Path"].apply(lambda s: os.path.join(data_path, "test", s + ".jpg"))
+
+    return train_df, train_df_fold, test_df
+
+
+if __name__ == "__main__":
+    data_path = "./data/"  # The path of the training and testing data.
+    save_path = "./pawpularity-result/result1/result"  # The path of saving the model.
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity.py:46]: consider to change the name of the save_path based on the name of the backbone + whether to use fusion or not.

File: examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity_submit.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    test_df["Image Path"] = test_df["Image Path"].apply(lambda s: os.path.join(data_path, "test", s + ".jpg"))
+    return train_df, test_df
+
+
+if __name__ == "__main__":
+    data_path = "../input/petfinder-pawpularity-score/"
+    train_df, test_df = load_data(data_path)
+
+    submission = pd.read_csv("../input/petfinder-pawpularity-score/sample_submission.csv")
+    save_paths = ["../input/pawpularity-automm-result/result1/result",
```
Comment (sxjscience) [inline examples/automm/Kaggle-Pawpularity/kaggle_Pawpularity_submit.py:32]: Consider to change the name of the result path.

**Comment by sxjscience** [comment]
Also, consider to change the filenames from `kaggle_Pawpularity_*` to `kaggle_pawpularity_*`.

**Review by sxjscience — APPROVED** [review]
Nice contribution!! Thanks!!!

**Comment by sxjscience** [comment]
We may change this line to not save to temporary directory: https://github.com/awslabs/autogluon/blob/d571bd2c72df90a9a5b42377a7189a3db0257785/multimodal/tests/unittests/test_predictor.py#L846
